File: app/src/main/java/com/botijasoftware/utils/Rectangle.java
package com.botijasoftware.utils;

public class Rectangle {

	public Rectangle() {

		X = 0;
		Y = 0;
		width = 0;
		height = 0;
	}
	
	public Rectangle(int px, int py, int w, int h) {

			X = px;
			Y = py;
			width= w;
			height = h;
	}
	
	public Rectangle(Vector2 position, Vector2 size) {

			X = (int) position.X;
			Y = (int) position.Y;
			width= (int) size.X;
			height = (int) size.Y;
	}

    public Rectangle(Rectangle rect) {

        X = rect.X;
        Y = rect.Y;
        width = rect.width;
        height = rect.height;
    }
	
	public Rectangle(Rectanglef rect) {

		X = (int) rect.X;
		Y = (int) rect.Y;
		width= (int) rect.width;
		height = (int) rect.height;
    }
	
	public void setSize(int w, int h) {
		width = w;
		height = h;
	}
	
	public void setSize(Vector2 v) {
		width = (int) v.X;
		height = (int) v.Y;
	}
	
	public void setPosition(int x, int y) {
		X = x;
		Y = y;
	}
	
	public void setPosition(Vector2 v) {
		X = (int) v.X;
		Y = (int) v.Y;
	}
	
	public void move(int incx, int incy) {
		X += incx;
		Y += incy;
	}
	
	public void move(Vector2 v) {
		X += (int)v.X;
		Y += (int)v.Y;
	}
	
	public boolean contains(int x, int y) {
		return ((x >= X) && ( x <= X + width) && (y >= Y) && (y <= Y + height)); 
	}
	
	public boolean contains(Vector2 v) {
		return (((int)v.X >= X) && ( (int)v.X <= X + width) && ((int)v.Y >= Y) && ((int)v.Y <= Y + height)); 
	}
	
	public void setCenter(int x, int y) {
		X = x - (width / 2);
		Y = y - (height / 2);
	}
	
	public void setCenter(Vector2 v) {
		X = (int)v.X - (width / 2);
		Y = (int)v.Y - (height / 2);
	}
	
	public Vector2 getCenter() {
		return new Vector2( X + (width / 2.0f), Y + (height / 2.0f));
	}
	
	public Vector2 getTopLeft() {
		return new Vector2( X, Y );
	}
	
	public Vector2 getTopRight() {
		return new Vector2( X + width, Y );
	}
	
	public Vector2 getBottomLeft() {
		return new Vector2( X, Y - height);
	}
	
	public Vector2 getBottomRight() {
		return new Vector2( X + width, Y - height);
	}

    public Rectangle clone() {
        return new Rectangle(X, Y, width, height);
    }
	
	public int X; //top left
	public int Y;
	public int width;
	public int height;

}

File: app/src/main/java/com/botijasoftware/utils/Rectangle.kt
package com.botijasoftware.utils

class Rectangle {

    constructor() {

        X = 0
        Y = 0
        width = 0
        height = 0
    }

    constructor(px: Int, py: Int, w: Int, h: Int) {

        X = px
        Y = py
        width = w
        height = h
    }

    constructor(position: Vector2, size: Vector2) {

        X = position.X.toInt()
        Y = position.Y.toInt()
        width = size.X.toInt()
        height = size.Y.toInt()
    }

    constructor(rect: Rectangle) {

        X = rect.X
        Y = rect.Y
        width = rect.width
        height = rect.height
    }

    constructor(rect: Rectanglef) {

        X = rect.X.toInt()
        Y = rect.Y.toInt()
        width = rect.width.toInt()
        height = rect.height.toInt()
    }

    fun setSize(w: Int, h: Int) {
        width = w
        height = h
    }

    fun setSize(v: Vector2) {
        width = v.X.toInt()
        height = v.Y.toInt()
    }

    fun setPosition(x: Int, y: Int) {
        X = x
        Y = y
    }

    fun setPosition(v: Vector2) {
        X = v.X.toInt()
        Y = v.Y.toInt()
    }

    fun move(incx: Int, incy: Int) {
        X += incx
        Y += incy
    }

    fun move(v: Vector2) {
        X += v.X.toInt()
        Y += v.Y.toInt()
    }

    fun contains(x: Int, y: Int): Boolean {
        return x >= X && x <= X + width && y >= Y && y <= Y + height
    }

    operator fun contains(v: Vector2): Boolean {
        return v.X.toInt() >= X && v.X.toInt() <= X + width && v.Y.toInt() >= Y && v.Y.toInt() <= Y + height
    }

    fun setCenter(x: Int, y: Int) {
        X = x - width / 2
        Y = y - height / 2
    }

    var center: Vector2
        get() = Vector2(X + width / 2.0f, Y + height / 2.0f)
        set(v) {
            X = v.X.toInt() - width / 2
            Y = v.Y.toInt() - height / 2
        }

    val topLeft: Vector2
        get() = Vector2(X.toFloat(), Y.toFloat())

    val topRight: Vector2
        get() = Vector2((X + width).toFloat(), Y.toFloat())

    val bottomLeft: Vector2
        get() = Vector2(X.toFloat(), (Y - height).toFloat())

    val bottomRight: Vector2
        get() = Vector2((X + width).toFloat(), (Y - height).toFloat())

    fun clone(): Rectangle {
        return Rectangle(X, Y, width, height)
    }

    var X: Int = 0 //top left
    var Y: Int = 0
    var width: Int = 0
    var height: Int = 0

}
